src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -2225,6 +2227,138 @@ class JNILocalRootsClosure : public OopClosure {
   virtual void do_oop(narrowOop* obj_p) { ShouldNotReachHere(); }
 };
 
+// Helper class to collect/report stack roots.
+class StackRootCollector {
+private:
+  JvmtiTagMap* _tag_map;
+  JNILocalRootsClosure* _blk;
+  // java_thread is needed only to report JNI local on top native frame;
+  // I.e. it's required only for platform/carrier threads or mounted virtual threads.
+  JavaThread* _java_thread;
+
+  oop _threadObj;
+  jlong _thread_tag;
+  jlong _tid;
+
+  bool _is_top_frame;
+  int _depth;
+  frame* _last_entry_frame;
+
+  bool report_stack_refs(StackValueCollection* values, jmethodID method, jlocation bci, jint slot_offset);
+
+public:
+  StackRootCollector(JvmtiTagMap* tag_map, JNILocalRootsClosure* blk, JavaThread* java_thread)
+    : _tag_map(tag_map), _blk(blk), _java_thread(java_thread),
+      _threadObj(nullptr), _thread_tag(0), _tid(0),
+      _is_top_frame(true), _depth(0), _last_entry_frame(nullptr)
+  {
+  }
+
+  bool set_thread(oop o);
+  // sets the thread and reports the reference to it with the specified kind.
+  bool set_thread(jvmtiHeapReferenceKind kind, oop o);
+
+  bool do_frame(vframe* vf);
+};
+
+bool StackRootCollector::set_thread(oop o) {
+  _threadObj = o;
+  _thread_tag = tag_for(_tag_map, _threadObj);
+  _tid = java_lang_Thread::thread_id(_threadObj);
+
+  _is_top_frame = true;
+  _depth = 0;
+  _last_entry_frame = nullptr;
+
+  return true;
+}
+
+bool StackRootCollector::set_thread(jvmtiHeapReferenceKind kind, oop o) {
+  return set_thread(o)
+         && CallbackInvoker::report_simple_root(kind, _threadObj);
+}
+
+bool StackRootCollector::report_stack_refs(StackValueCollection* values, jmethodID method, jlocation bci, jint slot_offset) {
+  for (int index = 0; index < values->size(); index++) {
+    if (values->at(index)->type() == T_OBJECT) {
+      oop o = values->obj_at(index)();
```

Comment:
I'd suggest to get rid of one-letter identifier like `o` and `c`.
They variables can be renamed to `obj` and `cont` instead.
It'd better to rename `slot_offset` to `offset`.

Reply:
changed variable names.
I think "offset" is not good name here, it's unclear what the offset is.
slot_offset shows that the offset is for reported slot parameter